### PR TITLE
lib: adding support for modem to updatehub library

### LINKED ISSUE #25704: lib: updatehub: Corrupted updated when receiving CoAP duplicate packages
The updatehub library is not properly handling the case where the CoAP packages are duplicated and we end flashing a broken firmware on the device's storage.

The short term solution is to have an engine that handles these packages reception and ensures we handle the proper sequence of blocks.

### PR DESCRIPTION
**Pull request by mtahirbutt**
This pull request addresses the support for slower air interfaces (modem for example) to updatehub. Originally updatehub was designed for high speed fixed LAN interfaces. Therefore, this implementation has been tested with DIGI XBC-M5-UT-001 XBee Cellular 3G modem with STM32L475 MCU BSP. It is expected that it should work with other interfaces as well. Some feedback is requested in this regard also. 
Additionally there is also a bug fix for #24853.

### CODE REVIEW

**Comment by nandojve** [comment]
Hi @mtahirbutt, thank you for sharing this. Now is far more clear.

CC @otavio

**Comment by mtahirbutt** [comment]
> 
> 
> Hi @mtahirbutt, thank you for sharing this. Now is far more clear.
> 
> CC @otavio

@nandojve the PR is ready for review, plz also add second reviewer. 

thanks.

File: lib/updatehub/updatehub.c [hunk 1]
```diff
@@ -54,6 +54,74 @@ LOG_MODULE_REGISTER(updatehub);
 #define UPDATEHUB_SERVER "coap.updatehub.io"
 #endif
 
+/* please refer to coap.c */
+
+#define GET_NUM_2(v) ((v) >> 4)
+
+static int attempts_download;
```
Comment (jukkar) [inline lib/updatehub/updatehub.c:61]: Could we place these newly introduced variable here and below to `ctx` struct?

File: lib/updatehub/updatehub.c [hunk 2]
```diff
@@ -137,7 +205,6 @@ static bool
 is_compatible_hardware(struct resp_probe_some_boards *metadata_some_boards)
 {
 	int i;
-
```
Comment (jukkar) [inline lib/updatehub/updatehub.c:140]: unrelated change, please remove. Having empty line here is good, so lets keep it.

File: lib/updatehub/updatehub.c [hunk 3]
```diff
@@ -383,13 +452,40 @@ static void install_update_cb(void)
 		LOG_ERR("Could not receive data");
 		goto cleanup;
 	}
-
```
Comment (jukkar) [inline lib/updatehub/updatehub.c:386]: Having empty line here is good, please remove this change.

File: lib/updatehub/updatehub.c [hunk 4]
```diff
@@ -383,13 +452,40 @@ static void install_update_cb(void)
...
 		ctx.code_status = UPDATEHUB_DOWNLOAD_ERROR;
 		goto cleanup;
 	}
 
+	/* check for the duplicate packet based upon block number if
+	 * duplicate discard the packet otherwise process it and
+	 * adjust the timers
+	 */
+
+	int block_num = get_num_block(&response_packet);
```
Comment (jukkar) [inline lib/updatehub/updatehub.c:466]: Please introduce variables at the beginning of the block.

File: lib/updatehub/updatehub.c [hunk 5]
```diff
@@ -383,13 +452,40 @@ static void install_update_cb(void)
...
+	/*  calculate the block number of coap data packet  */
+	if ((response_packet.max_len - response_packet.offset) <= 0
+		|| (block_num < 0)) {
+	/* ignore the rubbish packets */
```
Comment (jukkar) [inline lib/updatehub/updatehub.c:470]: Wrong indentation

File: lib/updatehub/updatehub.c [hunk 6]
```diff
@@ -383,13 +452,40 @@ static void install_update_cb(void)
...
+		LOG_ERR("Invalid data received or block number is < 0");
+		ctx.code_status = UPDATEHUB_OK;
+		goto cleanup;
+	}
+	if (block_num == expected_block_num) {
```
Comment (jukkar) [inline lib/updatehub/updatehub.c:475]: Please place empty line after ending } (apply to here and other places in your commit)

File: lib/updatehub/updatehub.c [hunk 7]
```diff
@@ -400,14 +496,32 @@ static void install_update_cb(void)
...
-	if (flash_img_buffered_write(&ctx.flash_ctx,
-				     response_packet.data + response_packet.offset,
-				     response_packet.max_len - response_packet.offset,
-				     ctx.downloaded_size == ctx.block.total_size) < 0) {
-		LOG_ERR("Error to write on the flash");
-		ctx.code_status = UPDATEHUB_INSTALL_ERROR;
-		goto cleanup;
+	if ((response_packet.max_len - response_packet.offset) <
+		CONFIG_IMG_BLOCK_BUF_SIZE) {
+	/* fix for zephyr-2.3-rc1 for STM32 series MCU as size of data write
```
Comment (jukkar) [inline lib/updatehub/updatehub.c:501]: Indentation wrong

File: lib/updatehub/updatehub.c [hunk 8]
```diff
@@ -429,17 +543,16 @@ static void install_update_cb(void)
 			goto cleanup;
 		}
 	}
-
```
Comment (jukkar) [inline lib/updatehub/updatehub.c:432]: Having here empty lines is fine, why are you removing these

**Review by jukkar — CHANGES_REQUESTED** [review]
Please check the code for style issues, there were lot of them. Also do not remove empty lines unnecessarily especially when they are not related to the actual functionality you are introducing.

File: lib/updatehub/updatehub.c (hunk 6 above)
Comment (mtahirbutt) [reply lib/updatehub/updatehub.c:480]: added spaces as you mentioned

File: lib/updatehub/updatehub.c (hunk 1 above)
Comment (mtahirbutt) [reply lib/updatehub/updatehub.c]: moved

File: lib/updatehub/updatehub.c (hunk 2 above)
Comment (mtahirbutt) [reply lib/updatehub/updatehub.c]: space added as suggested.

File: lib/updatehub/updatehub.c (hunk 3 above)
Comment (mtahirbutt) [reply lib/updatehub/updatehub.c]: done

File: lib/updatehub/updatehub.c (hunk 4 above)
Comment (mtahirbutt) [reply lib/updatehub/updatehub.c]: done

File: lib/updatehub/updatehub.c (hunk 5 above)
Comment (mtahirbutt) [reply lib/updatehub/updatehub.c:474]: corrected

File: lib/updatehub/updatehub.c [hunk 9]
```diff
@@ -402,9 +503,9 @@ static void install_update_cb(void)
 	}
 
 	if (flash_img_buffered_write(&ctx.flash_ctx,
-				     response_packet.data + response_packet.offset,
-				     response_packet.max_len - response_packet.offset,
-				     ctx.downloaded_size == ctx.block.total_size) < 0) {
+			    response_packet.data + response_packet.offset,
+			    response_packet.max_len - response_packet.offset,
+			    ctx.downloaded_size == ctx.block.total_size) < 0) {
```
Comment (mtahirbutt) [inline lib/updatehub/updatehub.c:508]: with new code, the previous places generates long line length warnings.

**Comment by jukkar** [comment]
Thanks for the fixes, you can squash the fix commit to the first one and resubmit branch.

**Comment by mtahirbutt** [comment]
> Thanks for the fixes, you can squash the fix commit to the first one and resubmit branch.

thanks for the message. done as suggested.

**Comment by nandojve** [comment]
We would like thank you @mtahirbutt by this contribution. It helps us to improve UpdateHub fixing 3 bugs, avoid unnecessary flash write, improve CoAP lib, and add support to slow wireless interfaces like modem.

This works was addressed on #25603, #25683 and #26093.
